pkg/runner/test_runner.go

```diff
@@ -6,282 +6,778 @@
+				StepName: ss.sb.TestStepLabel,
+				Target:   tgt.ID,
+			}
+		}
+		if ts.resCh == nil {
+			// This should not happen, must be an internal error.
+			return nil, fmt.Errorf("%s: target runner %s is not there, dropping result on the floor", ss, ts)
+		}
+		ss.log.Debugf("%s: result for %s: %v", ss, ts, res)
+		return ts.resCh, nil
+	}()
+	if err != nil {
+		return err
 	}
+	select {
+	case resCh <- res:
+		break
+	case <-ctx.Done():
+		break
+	}
+	return nil
 }
 
-// NewTestRunnerWithTimeouts initializes and returns a new TestRunner object with
-// custom timeouts
-func NewTestRunnerWithTimeouts(timeouts TestRunnerTimeouts) TestRunner {
-	return TestRunner{timeouts: timeouts}
+func (tr *testRunner) safeCloseOutCh(ss *stepState) {
+	defer func() {
+		if r := recover(); r != nil {
+			tr.mu.Lock()
+			ss.runErr = &cerrors.ErrTestStepClosedChannels{StepName: ss.sb.TestStepLabel}
+			tr.mu.Unlock()
+		}
+	}()
+	close(ss.outCh)
 }
 
-// State is a structure that models the current state of the test runner
-type State struct {
-	completedSteps   map[string]error
-	completedRouting map[string]error
-	completedTargets map[*target.Target]error
+// safeCloseErrCh closes error channel safely, even if it has already been closed.
+func (tr *testRunner) safeCloseErrCh(ss *stepState) {
+	defer func() {
+		if r := recover(); r != nil {
+			tr.mu.Lock()
+			ss.runErr = &cerrors.ErrTestStepClosedChannels{StepName: ss.sb.TestStepLabel}
+			tr.mu.Unlock()
+		}
+	}()
+	close(ss.errCh)
 }
 
-// NewState initializes a State object.
-func NewState() *State {
-	r := State{}
-	r.completedSteps = make(map[string]error)
-	r.completedRouting = make(map[string]error)
-	r.completedTargets = make(map[*target.Target]error)
-	return &r
+// stepReader receives results from the step's output channel and forwards them to the appropriate target runners.
+func (tr *testRunner) stepReader(ctx statectx.Context, ss *stepState) {
+	ss.log.Debugf("%s: step reader active", ss)
+	var err error
+	outCh := ss.outCh
+loop:
+	for {
+		select {
+		case tgt, ok := <-outCh:
+			if !ok {
+				ss.log.Debugf("%s: out chan closed", ss)
+				// At this point we may still have an error to report,
+				// wait until error channel is emptied too.
+				outCh = make(chan *target.Target)
+				tr.safeCloseErrCh(ss)
+				continue loop
+			}
+			if err = tr.reportTargetResult(ctx, ss, tgt, nil); err != nil {
+				break loop
+			}
+		case res, ok := <-ss.errCh:
+			if !ok {
+				ss.log.Debugf("%s: err chan closed", ss)
+				break loop
+			}
+			if err = tr.reportTargetResult(ctx, ss, res.Target, res.Err); err != nil {
+				break loop
+			}
+		case <-ctx.Done():
+			ss.log.Debugf("%s: canceled 3, draining", ss)
+			for {
+				select {
+				case <-outCh:
+					break loop
+				case <-ss.errCh:
+					break loop
+				case <-time.After(tr.shutdownTimeout):
+					tr.mu.Lock()
+					if ss.runErr == nil {
+						ss.runErr = &cerrors.ErrTestStepsNeverReturned{}
+					}
+					tr.mu.Unlock()
```

Comment:
So we timed-out but not interrupted the loop? Is this correct?

Reply:
we are interrupting the loop - just below this we forcibly close the output channel even though the run method is still running precisely to release the reader. the reader will close error channel and exit, and it will also release the goroutine we started above. as a result, the only leak will be of the goroutine stuck in the step's Run method (this is verified in the test using CheckLeakedGoRoutines)